torch_geometric/data/datapipes.py

```diff
@@ -64,6 +73,21 @@ def __iter__(self):
             yield data
 
 
+class DatasetAdapter(IterDataPipe):
```

Comment:
I'm not sure if it would make more sense to have this inherit from `MapDataPipe` since that is more natural for PyG's Dataset implementations.

Would also like to somehow lazily initialise the `len` to support sharding.